File: src/main/java/com/machfour/macros/core/Table.java
package com.machfour.macros.core;

import java.util.List;
import java.util.Map;

public interface Table<M> {
    String getName();

    List<Column<M, ?>> getColumns();
    // return all FK columns
    List<Column.Fk<M, ?, ?>> getFkColumns();

    Map<String, Column<M, ?>> getColumnsByName();

    Column<M, ?> getColumnForName(String name);

    Column<M, Long> getIdColumn();

    Column<M, Long> getCreateTimeColumn();

    Column<M, Long> getModifyTimeColumn();

    // returns a list of columns that can be used to identify an individual row,
    // if such a list exists for this table. If not, an empty list is returned.
    List<Column<M, ?>> getSecondaryKeyCols();
    // special case when secondary key has a single column.
    Column<M, ?> getNaturalKeyColumn();

    Factory<M> getFactory();

    default M construct(ColumnData<M> dataMap, ObjectSource objectSource) {
        return getFactory().construct(dataMap, objectSource);
    }

}

File: src/main/java/com/machfour/macros/sql/Table.kt
package com.machfour.macros.sql

import com.machfour.macros.orm.Factory
import com.machfour.macros.orm.ObjectSource

interface Table<M> {
    val name: String
    val columns: List<Column<M, *>>

    // return all FK columns
    val fkColumns: List<Column.Fk<M, *, *>>
    val columnsByName: Map<String, Column<M, *>>

    val idColumn: Column<M, Long>
    val createTimeColumn: Column<M, Long>
    val modifyTimeColumn: Column<M, Long>

    // returns a list of columns that can be used to identify an individual row,
    // if such a list exists for this table. If not, an empty list is returned.
    val secondaryKeyCols: List<Column<M, *>>

    // special case when secondary key has a single column.
    val naturalKeyColumn: Column<M, *>?
    val factory: Factory<M>

    fun construct(dataMap: RowData<M>, objectSource: ObjectSource): M = factory.construct(dataMap, objectSource)
}